File: runtime/decision_engine.py
"""Smart decision engine for agent selection and routing."""
from __future__ import annotations

import os
import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

from hooks._agent_registry import detect_available_models, AGENT_REGISTRY


class TaskComplexity(Enum):
    TRIVIAL = 1
    SIMPLE = 2
    MODERATE = 3
    COMPLEX = 4
    EXTREME = 5


@dataclass
class DecisionContext:
    prompt: str
    project_dir: str
    domain_hints: list[str] = field(default_factory=list)
    intent: str | None = None
    history: list[dict] = field(default_factory=list)


@dataclass
class AgentRecommendation:
    agent_name: str
    category: str
    provider: str
    confidence: float
    fallback: list[str] = field(default_factory=list)
    reasoning: str


COMPLEXITY_INDICATORS = {
    TaskComplexity.TRIVIAL: [
        r"typo", r"fix.*spelling", r"rename.*file", r"change.*label",
    ],
    TaskComplexity.SIMPLE: [
        r"simple", r"quick", r"minor", r"small", r"add.*comment",
    ],
    TaskComplexity.MODERATE: [
        r"implement", r"add.*feature", r"refactor", r"fix.*bug",
    ],
    TaskComplexity.COMPLEX: [
        r"design.*system", r"architecture", r"multi.*agent", r"parallel",
        r"security.*audit", r"performance.*optim",
    ],
    TaskComplexity.EXTREME: [
        r"redesign", r"rewrite.*entire", r"machine.*learning",
        r"novel.*algorithm", r"research.*new",
    ],
}

DOMAIN_AGENT_MAP = {
    frozenset(["ui", "ux", "css", "visual", "design", "frontend", "component"]): "frontend-designer",
    frozenset(["api", "rest", "graphql", "endpoint", "contract"]): "api-builder",
    frozenset(["security", "auth", "vulnerability", "audit"]): "security-auditor",
    frozenset(["database", "sql", "migration", "schema"]): "database-engineer",
    frozenset(["test", "spec", "coverage", "e2e"]): "testing-engineer",
    frozenset(["deploy", "docker", "kubernetes", "ci", "cd", "infra"]): "infra-engineer",
    frozenset(["release", "version", "publish", "changelog"]): "release-engineer",
}


def analyze_complexity(prompt: str) -> TaskComplexity:
    prompt_lower = prompt.lower()
    for complexity, patterns in COMPLEXITY_INDICATORS.items():
        for pattern in patterns:
            if re.search(pattern, prompt_lower):
                return complexity
    return TaskComplexity.MODERATE


def extract_domain(prompt: str) -> str | None:
    prompt_lower = prompt.lower()
    words = set(re.findall(r'\b\w+\b', prompt_lower))

    for domain_words, agent_name in DOMAIN_AGENT_MAP.items():
        if domain_words & words:
            return agent_name

    return None


def get_provider_priority() -> list[str]:
    available = detect_available_models()
    priority = []

    if available.get('codex-cli'):
        priority.append('codex')
    if available.get('gemini-cli'):
        priority.append('gemini')
    if available.get('claude-code'):
        priority.append('claude')
    if available.get('claude'):
        priority.append('claude')
    if available.get('opencode-cli'):
        priority.append('opencode')
    if available.get('kimi-cli'):
        priority.append('kimi')
    if available.get('cursor-cli'):
        priority.append('cursor')

    return priority


def map_category_to_provider(category: str, available: dict[str, bool]) -> str:
    category_provider_map = {
        'visual-engineering': ['gemini-cli', 'claude'],
        'deep': ['codex-cli', 'claude'],
        'ultrabrain': ['codex-cli', 'claude'],
        'quick': ['claude', 'opencode-cli'],
    }

    preferred = category_provider_map.get(category, ['claude'])
    for p in preferred:
        if available.get(p):
            return p.replace('-cli', '')

    for p in ['codex', 'gemini', 'claude', 'opencode', 'kimi']:
        if available.get(f'{p}-cli') or available.get(p):
            return p

    return 'claude'


def decide_agent(context: DecisionContext) -> AgentRecommendation:
    complexity = analyze_complexity(context.prompt)
    domain_agent = extract_domain(context.prompt)

    if domain_agent:
        agent_config = AGENT_REGISTRY.get(domain_agent, {})
    else:
        agent_config = {}

    category = agent_config.get('task_category', 'unspecified-high')
    available = detect_available_models()
    provider = map_category_to_provider(category, available)

    confidence = 0.5
    if domain_agent:
        confidence += 0.3
    if complexity == TaskComplexity.TRIVIAL or complexity == TaskComplexity.SIMPLE:
        confidence += 0.1

    fallback = []
    if provider != 'claude':
        fallback.append('claude')
    if provider != 'codex' and available.get('codex-cli'):
        fallback.append('codex')
    if provider != 'gemini' and available.get('gemini-cli'):
        fallback.append('gemini')

    reasoning = f"Complexity={complexity.name}, Domain={domain_agent or 'general'}, Provider={provider}"

    return AgentRecommendation(
        agent_name=domain_agent or 'task',
        category=category,
        provider=provider,
        confidence=confidence,
        fallback=fallback,
        reasoning=reasoning,
    )


def get_fallback_chain(primary_failures: list[str]) -> list[str]:
    available = detect_available_models()
    chain = []

    order = ['claude', 'codex', 'gemini', 'opencode', 'kimi', 'cursor']
    for p in order:
        if p not in primary_failures:
            if available.get(f'{p}-cli') or available.get(p):
                chain.append(p)

    return chain


def should_parallelize(context: DecisionContext) -> bool:
    prompt_lower = context.prompt.lower()
    parallel_indicators = [
        r"parallel", r"concurrent", r"simultaneous",
        r"multiple.*files", r"batch", r"all.*at once",
    ]

    for pattern in parallel_indicators:
        if re.search(pattern, prompt_lower):
            return True

    if analyze_complexity(context.prompt) >= TaskComplexity.COMPLEX:
        return True

    return False

File: src/orchestration/decision-engine.ts
// Ported from runtime/decision_engine.py

import type {
  AgentRecommendation,
  TaskComplexity,
} from "../interfaces/orchestration.js";

// ---------------------------------------------------------------------------
// Complexity indicators — regex patterns per complexity level
// ---------------------------------------------------------------------------

const COMPLEXITY_INDICATORS: ReadonlyMap<TaskComplexity, readonly RegExp[]> =
  new Map<TaskComplexity, readonly RegExp[]>([
    [
      "trivial",
      [/typo/i, /fix\s*spelling/i, /rename\s*file/i, /change\s*label/i],
    ],
    [
      "simple",
      [/\bsimple\b/i, /\bquick\b/i, /\bminor\b/i, /\bsmall\b/i, /add\s*comment/i],
    ],
    [
      "moderate",
      [/\bimplement\b/i, /add\s*feature/i, /\brefactor\b/i, /fix\s*bug/i],
    ],
    [
      "complex",
      [
        /design\s*system/i,
        /\barchitecture\b/i,
        /multi\s*agent/i,
        /\bparallel\b/i,
        /security\s*audit/i,
        /performance\s*optim/i,
        /\bOAuth\b/i,
        /\bauth\s*system\b/i,
      ],
    ],
    [
      "extreme",
      [
        /\bredesign\b/i,
        /rewrite\s*entire/i,
        /machine\s*learning/i,
        /novel\s*algorithm/i,
        /research\s*new/i,
      ],
    ],
  ]);

// ---------------------------------------------------------------------------
// Domain → agent mapping
// ---------------------------------------------------------------------------

interface DomainEntry {
  readonly keywords: ReadonlySet<string>;
  readonly agent: string;
  readonly category: string;
}

const DOMAIN_AGENT_MAP: readonly DomainEntry[] = [
  {
    keywords: new Set([
      "ui",
      "ux",
      "css",
      "visual",
      "design",
      "frontend",
      "component",
    ]),
    agent: "frontend-designer",
    category: "visual-engineering",
  },
  {
    keywords: new Set(["api", "rest", "graphql", "endpoint", "contract"]),
    agent: "api-builder",
    category: "deep",
  },
  {
    keywords: new Set(["security", "auth", "vulnerability", "audit"]),
    agent: "security-auditor",
    category: "ultrabrain",
  },
  {
    keywords: new Set(["database", "sql", "migration", "schema"]),
    agent: "database-engineer",
    category: "deep",
  },
  {
    keywords: new Set(["test", "spec", "coverage", "e2e"]),
    agent: "testing-engineer",
    category: "quick",
  },
  {
    keywords: new Set(["deploy", "docker", "kubernetes", "ci", "cd", "infra"]),
    agent: "infra-engineer",
    category: "deep",
  },
  {
    keywords: new Set(["code", "codex", "implement", "build", "develop"]),
    agent: "codex",
    category: "deep",
  },
  {
    keywords: new Set(["research", "gemini", "analyze", "investigate"]),
    agent: "gemini",
    category: "ultrabrain",
  },
] as const;

// ---------------------------------------------------------------------------
// Provider priority by category
// ---------------------------------------------------------------------------

const CATEGORY_PROVIDER_MAP: Readonly<Record<string, readonly string[]>> = {
  "visual-engineering": ["gemini", "claude"],
  deep: ["codex", "claude"],
  ultrabrain: ["codex", "claude"],
  quick: ["claude", "opencode"],
};

const DEFAULT_PROVIDER = "claude";

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Falls back to `"moderate"` when no indicator pattern matches. */
export function scoreComplexity(task: string): TaskComplexity {
  for (const [complexity, patterns] of COMPLEXITY_INDICATORS) {
    for (const pattern of patterns) {
      if (pattern.test(task)) {
        return complexity;
      }
    }
  }
  return "moderate";
}

function extractDomain(
  prompt: string,
): { agent: string; category: string } | undefined {
  const words = new Set(prompt.toLowerCase().match(/\b\w+\b/g) ?? []);

  for (const entry of DOMAIN_AGENT_MAP) {
    for (const keyword of entry.keywords) {
      if (words.has(keyword)) {
        return { agent: entry.agent, category: entry.category };
      }
    }
  }
  return undefined;
}

function selectProvider(category: string): string {
  const candidates = CATEGORY_PROVIDER_MAP[category];
  if (candidates && candidates.length > 0) {
    return candidates[0] ?? DEFAULT_PROVIDER;
  }
  return DEFAULT_PROVIDER;
}

function buildFallback(primary: string): readonly string[] {
  const all = ["claude", "codex", "gemini", "opencode"];
  return all.filter((p) => p !== primary);
}

export function recommendAgent(task: string): AgentRecommendation {
  const complexity = scoreComplexity(task);
  const domain = extractDomain(task);

  const agentName = domain?.agent ?? "task";
  const category = domain?.category ?? "unspecified-high";
  const provider = selectProvider(category);

  let confidence = 0.5;
  if (domain) {
    confidence += 0.3;
  }
  if (complexity === "trivial" || complexity === "simple") {
    confidence += 0.1;
  }

  const fallback = buildFallback(provider);
  const reasoning = `Complexity=${complexity}, Domain=${agentName}, Provider=${provider}`;

  return {
    agentName,
    category,
    provider,
    confidence,
    fallback,
    reasoning,
    complexity,
  };
}
